### PR TITLE
308 fix avet upsert

### PR DESCRIPTION
**Pull request by grischoun**
Fixes issue with upserting on :avet.

### CODE REVIEW

**Comment by grischoun** [comment]
This fixes issue #308.

**Comment by grischoun** [comment]
I ran the benchmarks against this change and there is no change in performance compared to what we had before (current dev. branch).

**Review by jsmassa — DISMISSED** [review]
Nice work

**Comment by whilo** [comment]
I need to look into the details still, but have you made sure that the old `UpsertOp` values can still be read correctly by the read handlers? I guess we still don't have automatic upgrade tests.

**Comment by grischoun** [comment]
> I need to look into the details still, but have you made sure that the old `UpsertOp` values can still be read correctly by the read handlers? I guess we still don't have automatic upgrade tests.

Yes, read handlers are working. 
Old UpsertOps could end up with a nil value for the newly introduced field `indices` but this line protects against that: https://github.com/replikativ/datahike/blob/07b0ea4bd5576f13bd6bd08feff3ff4ad030fc3e/src/datahike/index/hitchhiker_tree/upsert.cljc#L10
This leads to the safe but slower path.
I think it is the best we can do for now. May be later with Wanderung, knowing on which kind of tree we are working on, we will be able to fix it by setting the correct value for `indices`.

**Comment by whilo** [comment]
Is there a default value that can be used on the old values being read? It is faster and cleaner to extend the read handler than to substitute values in the logic.

File: src/datahike/index/hitchhiker_tree/upsert.cljc [hunk 1]
```diff
@@ -2,54 +2,86 @@
...
+  [old-keys new indices]
+  (when (seq old-keys)
+    (let [mask (mask new indices)]
+      (when-let [candidates (subseq old-keys >= mask)]
         (->> candidates
              (map first)
-             (take-while #(and (= a (first %)) (= b (second %))))
-             reverse
-             first)))))
+             ((if (prefix? indices) take-while filter)
```
Comment (grischoun) [inline src/datahike/index/hitchhiker_tree/upsert.cljc:41]: What is going on here?
When the 'indices' in the vector are strictly increasing by 1, like in [0 1 2], then we know that we are dealing with results returned from 'subseq' which are sorted from the 1st, 2nd and third entries in the datom. Thus we can use 'take-while' to retrieve all matching datoms until there is no match.
But when the 'indices' vector is like [0 2], which is the case for :avet, 'subseq' won't return datoms sorted with 'a' then 'e' but rather with 'a' then 'v' and 'e'. Thus we can't use 'take-while' but are forced to filter out all the results from 'subseq' that match 'a' and 'e'.
In summary, 'prefix?' is used to check whether the 'indices' vector represents an index where the prefix of datoms are used to sort them out or not.

Comment (whilo) [reply src/datahike/index/hitchhiker_tree/upsert.cljc:41]: Ok, thank you for explaining!

Comment (whilo) [reply src/datahike/index/hitchhiker_tree/upsert.cljc:41]: Ok. Thank you for explaining!

File: src/datahike/index/hitchhiker_tree/upsert.cljc [hunk 2]
```diff
@@ -2,54 +2,86 @@
...
 (defn old-retracted
   "Returns a new datom to insert in the tree to signal the retraction of the old datom."
-  [kvs key]
-  (when-let [old (old-key kvs key)]
+  [kvs key indices]
+  (when-let [old (old-key kvs key indices)]
     (let [[a b c _] old
           [_ _ _ nt] key]
       ;; '-' means it is retracted and 'nt' is the current transaction time.
       [a b c (- nt)])))
```
Comment (whilo) [inline src/datahike/index/hitchhiker_tree/upsert.cljc:77]: Do we use negative transaction times for retractions in general?

**Review by whilo — APPROVED** [review]
Looks good to me.

File: src/datahike/index/hitchhiker_tree.cljc [hunk 3]
```diff
@@ -119,14 +119,21 @@
 (defn -insert [tree ^Datom datom index-type op-count]
   (hmsg/insert tree (datom->node datom index-type) nil op-count))
 
+(defn- index-type->indices [index-type]
+  (case index-type
+    :eavt [0 1]
+    :aevt [0 1]
+    :avet [0 2]
```
Comment (whilo) [inline src/datahike/index/hitchhiker_tree.cljc:126]: Is this true if we operate on the `avet` index itself? This is only true if we operate on `aevt`, right?

**Review by whilo — CHANGES_REQUESTED** [review]
Maybe you can clarify this before merging.

File: src/datahike/index/hitchhiker_tree/upsert.cljc (hunk 2 above)
Comment (grischoun) [reply src/datahike/index/hitchhiker_tree/upsert.cljc:77]: That's what I always thought. @kordano Can you confirm this?

File: src/datahike/index/hitchhiker_tree.cljc (hunk 3 above)
Comment (grischoun) [reply src/datahike/index/hitchhiker_tree.cljc:126]: @whilo Sorry, I don't understand the question.

Comment (whilo) [reply src/datahike/index/hitchhiker_tree.cljc:126]: I meant the only case with `[0 2]`, `:avet`. I guess we do not work on the actual `avet` index in this case, otherwise it would be `[0 1]`.

Comment (grischoun) [reply src/datahike/index/hitchhiker_tree.cljc:126]: Yes, we don’t work on avet.

Comment (whilo) [reply src/datahike/index/hitchhiker_tree.cljc:126]: Ok, can you explain why not? I would have guessed that we need to upsert in there as well(?). Sorry if this is a question with an obvious answer.

Comment (grischoun) [reply src/datahike/index/hitchhiker_tree.cljc:126]: Sorry my previous answer is based on a misunderstanding of your question. 
We *do* upsert on :avet but we need to use :a and :e, i.e., position 0 and 2 (and not 0 and 1) to correctly perform upserts on :avet.

**Comment by grischoun** [comment]
This PR also fixes #338 .

File: src/datahike/index/hitchhiker_tree/upsert.cljc [hunk 4]
```diff
@@ -58,28 +90,28 @@
...
-          (fn [{:keys [key value]}]
-            (map->UpsertOp {:key key :value value}))
+          (fn [{:keys [key value ts indices]}]
+            (map->UpsertOp {:key key :value value :ts ts :indices (or indices [0 1])}))
 
           'datahike.index.hitchhiker_tree.upsert.temporal-UpsertOp
-          (fn [{:keys [key value]}]
-            (map->temporal-UpsertOp {:key key :value value}))})
+          (fn [{:keys [key value ts indices]}]
+            (map->temporal-UpsertOp {:key key :value value :ts ts :indices (or indices [0 1])}))})
```
Comment (grischoun) [inline src/datahike/index/hitchhiker_tree/upsert.cljc:116]: These are the fixes for #338.

File: src/datahike/index/hitchhiker_tree.cljc (hunk 3 above)
Comment (whilo) [reply src/datahike/index/hitchhiker_tree.cljc:132]: Thanks, that makes sense.

**Review by whilo — APPROVED** [review]

**Review by whilo — CHANGES_REQUESTED** [review]
I think it is important to either go with `op-count` everywhere or `ts`, but not to have both. As far as I see here this is still the case for `UpsertOp` and `temporal-UpsertOp`. I think you said that you are still fixing this, I am just making sure that my review is uptodate and we can merge as soon as it is ready.

**Comment by grischoun** [comment]
> I think it is important to either go with `op-count` everywhere or `ts`, but not to have both. As far as I see here this is still the case for `UpsertOp` and `temporal-UpsertOp`. I think you said that you are still fixing this, I am just making sure that my review is uptodate and we can merge as soon as it is ready.


I don't think we can rewrite everything to `op-count` (or `ts`) for now because there might be existing databases that have been serialised with `ts`. Once we have 'Wandering' though, we could migrate these databases and get rid of `ts`.

**Comment by grischoun** [comment]
'ts' is replaced by 'op-count' now. 
Ready for another round of reviews.

**Review by whilo — APPROVED** [review]
This looks good to me.
